### PR TITLE
Optimization

### PR DESCRIPTION
**Pull request by ferrine**
Bayesian optimization over posterior latent space is an interesting sort of problem that becomes real with this PR. Notebook with toy example is provided

@fonnesbeck you can find Histogram application for SVGD there

### CODE REVIEW

**Comment by ferrine** [comment]
@twiecki @fonnesbeck I think I need a review. Do we need bayesian optimization in PyMC3?

**Comment by springcoil** [comment]
Just a comment - and adding my two cents here. I've got no moral objection to having Bayesian Optimization in PyMC3, if you're willing to maintain the code. I'll give it a quick review now.

File: pymc3/optimization.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        self.hist = np.asarray(())
+
+    @change_flags(compute_test_value='off')
+    def refresh(self, kwargs=None):
+        """
+        Recompile step_function and reset updates
+
+        Parameters
+        ----------
+        kwargs : kwargs for theano.function
```
Comment (springcoil) [inline pymc3/optimization.py]: Hmm this could do with a bit more fleshing out in the docstring I think. Maybe with an example.

Comment (springcoil) [inline pymc3/optimization.py]: Hmm this could do with a bit more fleshing out in the docstring I think. Maybe with an example.

**Comment by fonnesbeck** [comment]
I think this could be useful for fitting GP models that include discrete parameters, so I'm in favor.

**Comment by springcoil** [comment]
Does anyone know any good papers to read on this? Would one of the application areas be something like Bandit problems?

**Comment by ferrine** [comment]
@springcoil You can see [this](https://github.com/pymc-devs/pymc3/pull/1904#issuecomment-286988104) message from PR thread, I've described some applications

**Comment by twiecki** [comment]
So this replaces sample_ppc?

**Comment by ferrine** [comment]
@twiecki how does it replace?

File: pymc3/optimization.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        """
+        updates = self.optimizer(self.loss, self.wrt)
+        self.step_function = theano.function([], self.loss, updates=updates, **kwargs)
+        self.hist = np.asarray(())
+
+    def fit(self, n=5000, callbacks=()):
+        """
+        Perform optimization steps
+        Parameters
```
Comment (twiecki) [inline pymc3/optimization.py]: Needs new-line I think before Parameters.

**Comment by twiecki** [comment]
@ferrine the notebook needs way more descriptions, it's not clear to me at all. what are you trying to optimize, where could this be useful etc. For example: `opt = pm.Optimizer(approx, y.sum(), [x], optimizer=sgd)` Why do you sum y?

**Comment by fonnesbeck** [comment]
This could use some actual documentation (not just docstrings), that is, adding to the files in `pymc3/docs` so that there is a guide to usage. At least open a PR for the docs, if not here.

See #1968

**Comment by springcoil** [comment]
Yeah I'd like to encourage actual documentation for this as well.

**Comment by ferrine** [comment]
I've updated notebook. Hope it reveals the purpose of this tool better

**Comment by twiecki** [comment]
@ferrine Much better, this is really neat.

* I would describe what `histogram.apply_replacements(y, deterministic=True).eval()` does.
* So Histogram is like a trace with special properties? Does it also work for ADVI? What about NUTS?
* `from itertools import zip_longest` is not used.
* `y_ = abc[0]*x_**2  + abc[1]*x_ + abc[2]` can't you use `f()` you defined above instead?
* Quite a few typos: Condider, bayesian -> Bayesian

**Comment by ferrine** [comment]
Okay, I'll take it all in account and change notebook soon

**Comment by ferrine** [comment]
Done

**Comment by twiecki** [comment]
@ferrine The docs are really enlightening. I wonder if `Histogram` is the right name. What about `Posterior`?

File: pymc3/optimization.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+__all__ = [
+    'Optimizer'
+]
+
+
+class Optimizer(object):
+    """
+    Optimization with posterior replacements
+    Parameters
```
Comment (twiecki) [inline pymc3/optimization.py]: Also new-line.

**Comment by ferrine** [comment]
@twiecki I think `Posterior` does not reveal the nature of that class that is just storing particles. It can confuse user and if trace is bad it's not `Posterior` even.

**Comment by twiecki** [comment]
Maybe, but `Histogram` is certainly confusing already. What about that makes it a histogram? `PosteriorApprox`, `PosteriorSamples`, `PosteriorEst`?

Another question is if we could marry this with Trace somehow?

**Comment by ferrine** [comment]
@twiecki what about `Particles`?

**Comment by ferrine** [comment]
It is married with trace already, or what do you mean?

**Comment by twiecki** [comment]
`Particles` isn't bad, but I don't like the co-notation with particle samplers.

I mean to merge the functionality of Histogram into Trace. I don't necessarily think it's a good idea, but there are definitely parallels. As I understand it, Histogram = Trace + computational capabilities.

**Comment by ferrine** [comment]
We support different backends in Trace I'm not sure all of them are possible in this paradigm. That's why I only use common Trace interface to initialize Histogram and immediately forget about it

**Comment by twiecki** [comment]
I still think we should call it something with Posterior, because that's what the computations happen on.

**Comment by ferrine** [comment]
Maybe `Empirical`? Empirical distribution is the one that consists of samples only. I do not like `Posterior` there as I inherit from `Approximation` and `Posterior(Approximation)` seems to be wrong name from that point of view. `Empirical(Approximation)` is be more intuitive

**Comment by twiecki** [comment]
I see, what does Approximation do again? That's the base-class for VI?

**Comment by ferrine** [comment]
That's the base class for VI results. It can perform sampling

**Comment by twiecki** [comment]
I think it's time we think a bit deeper about the structure naming of `variational`. Could you help by providing an overview of the classes and their purpose and inheritance structure?

**Comment by ferrine** [comment]
Do you mean Developer Guide?

**Comment by twiecki** [comment]
Just a quick list in the flavor of:
* Histogram: Inherits from Approximation. Collects posterior samples and allows computations on them like ...
* Approximation: Store VI results, supports sampling from VI estimate.

This is just for us to think about good names for each part and how things fit together. We could extend this to a more complete developer guide later.

**Comment by ferrine** [comment]
I also think that soon it will be time to port all our inference results to Approximation and bind helper functions with it.

**Comment by springcoil** [comment]
Yeah a quick list or diagram would be useful @ferrine I'm really confused about this structure.

**Comment by ferrine** [comment]
OPVI Structure


* Operator
	- It is inference method, it takes Model, Approximation, TestTunction(if needed) and defines objective to minimize.
	- the only thing that is customized is `.apply(f)` it takes a test function and teturns loss. If not TF is needed it just returns loss that depends on `z`(accessible via `self.input`). For example KL's apply is just one line 
	
	```
	    def apply(self, f):
           z = self.input
           return self.logq_norm(z) - self.logp_norm(z)
	```
	- module: `variational.operators`
* TestFunction
	- Test Function is needed when Inference method requires something exotic to be applied to posterior samples. For instance Kernelized Stein Discrepancy requires a Kernel function with it's derivatives wrt particles. That Kernel Function can be treated as Test Function
 	- module: `variational.test_functions`
* ObjectiveFunction
	- Internally constructed with Operator and defines how to create updates, calculate loss, perform one training step
	- module: does not have any special module
* Approximation
 	- Stores results of inference. It provides `.random()` method and optionally `.logq(z)` if it is parametric. Every inherited class can do sampling. 
 	- Sampling is done in symbolic way so it is possible to replace original pymc3 Variables with their view from Approximation. That is done in `.apply_replacements(...)` and `.sample_node(...)`. They do exactly what you expect they should do. Constructed node does not depend any more on pymc3 Variables
	- module: `variational.approximations`
* Inference
	- Links together Model, Operator, Test Function and Approximation. creates and stores Objective Function. Has several utils for doing inference itself and profiling training stage.
	- module: `variational.inference`

**Comment by springcoil** [comment]
I think this could be stuck into the docs like that. Thanks :)

**Comment by ferrine** [comment]
I think this question should be resolved before v3.1

I don't like `.sample_vp` method name. In Histogram they are not variational at all

**Comment by twiecki** [comment]
@ferrine Which structure do you think we should move to? Seems like some things could be integrated into this.

**Comment by ferrine** [comment]
1. module is not for variational purposes in some cases
2. `sample_vp` is not only for variational parameters
3. Histogram is a bad name

So there are three points to think over. Edward has a good practice to store all inference methods at one place. Ideally is to do the same in pymc3, but so serious api changes are harmful.

If want to integrate Histogram to step methods we will meet architecture problems for sure in the future. General structure is not consistent. We have step methods, VI methods and they produce results with different interface. Moreover they are all stored in different modules.
structure like
```
inference/
    __init__.py # defines .fit, imports all methods
    base.py # defines 'Approximation', 'Histogram' classes
    step/
        methods.py # for step methods from step module
        ...
    variational/
        methods.py # for variational methods from variational module
        opvi.py # like base for variational 
        ... 
        
```

**Comment by ferrine** [comment]
Idea: add Lagrange method here
See https://en.m.wikipedia.org/wiki/Lagrange_multiplier

It is done in one line of code

**Comment by ferrine** [comment]
I got lots of ideas for optimization module. It will be a great tool for practical problems

**Comment by twiecki** [comment]
@taku-y If you have time would also be interested in your perspective.

**Comment by taku-y** [comment]
@twiecki I'm really sorry for not being responsive. Basically, I totally agrees this PR. Here's are short comments:

1. The OPVI framework looks fine to me. It can describe a variety of inference algorithm and models in a consistent way, although currently not consistent with sampling methods.
2. It might better to show `sample_vp` not always for variational posterior. The name can be changed, although we need to take care of backward compatibility.
3. I vote `Empirical` instead of `Histogram`.
4. Bayesian optimization is definitely useful.

**Comment by fonnesbeck** [comment]
Maybe `sample_approx` to replace `sample_vp`, since they will all be approximations we are attempting to sample from, correct?

**Comment by twiecki** [comment]
Another idea for a better name of `Histogram` that @jsalvatier had is `EmpiricalDistribution`, or alternatively, `ApproxDistribution`. I kinda like the latter and it goes into a similar direction as what @fonnesbeck just said. Then `sample_approx()` would internally call `ApproxDistribution.sample()`.

**Comment by twiecki** [comment]
I also like @ferrine's proposed restructuring:
```
inference/
    __init__.py # defines .fit, imports all methods
    base.py # defines 'Approximation', 'Histogram' classes
    step/
        methods.py # for step methods from step module
        ...
    variational/
        methods.py # for variational methods from variational module
        opvi.py # like base for variational 
        ... 
```

I don't think we should shy away from changing the API / structure of the code to unify MCMC and VI interfaces a bit.

**Comment by twiecki** [comment]
@ferrine Above you write for the `Approximate` class: 
> Stores results of inference. It provides .random() method and optionally .logq(z) if it is parametric. Every inherited class can do sampling.

How is that different from `Histogram`, couldn't `Approximate` handle that too?

**Comment by ferrine** [comment]
@twiecki `Approximation` is a very broad name, best for base class.
`Histogram` seems to be a final class, that is used later for VI, it also can be used for storing step methods results. It also has some specific methods and different sampling idea. So the best option is inheriting it form Approximation in the root directory of module.

**Comment by twiecki** [comment]
@ferrine So `Approximation` is only for VI results? What else inherits from `Approximation`?

**Comment by ferrine** [comment]
@twiecki `Approximation` itself is *mainly* for VI but some changes are done for `Histogram`(that's why *mainly*). 
Ideally they both should inherent from some base class. But when implementing `Histogram` I saw no profit from it and got what we have.

**Comment by twiecki** [comment]
I don't really see the purpose of inheritance then. And if it's purely a base class, could do `ApproxBase` and `ApproxDistribution(ApproxBase)` which would also make things more obvious. But I think we should merge it into one.

**Comment by ferrine** [comment]
Inheritance is for writing less code for more functionality and isinstance checks. I think I don't clearly understand what is `ApproxDistribution`.

**Comment by twiecki** [comment]
> Inheritance is for writing less code for more functionality and isinstance checks.

I don't see how that's the case if there's only a single inherited class and an abstract base class. It just makes things more complicated.

`ApproxDistribution` is any approximated (either through samples (MCMC) or parameterized (VI)) distribution. In this case the posterior. You can sample from it etc. Actually thinking about it that way it should have similar API to the existing probability distributions we already have, no?

**Comment by ferrine** [comment]
> I don't see how that's the case if there's only a single inherited class and an abstract base class. It just makes things more complicated.

Writing Histogram without Inheritance is bad idea too. I've already avoided unnecessary Inheritance. I don't have any idea how to make code simpler remaining functionality. There are a lot of common parts.

**Comment by twiecki** [comment]
@ferrine I looked at the code again and `Histogram` is not the only class inheriting from `Approximate`, there's also `MeanField` and `FullRank` in which case inheritance makes sense. Which functionality does `Histogram` use from `Approximate`?

**Comment by jsalvatier** [comment]
Very interesting! 

If this is for Bayesian optimization, why is there so much variational specific stuff? You do Bayesian optimization with other methods. Oh, I see this is why Thomas is talking about merging the MCMC and variational representations. 

What exactly is Histogram's role? If I'm reading this correctly, is the symbolic representation of the posterior (in this case stored as samples). You need that so that you can evaluate the model at specific points to compute the expected loss. 

For bayesian optimization are you always going to go through sampling? My guess is yes. 

If so, then you just need a common representation for a posterior distribution that can be used to generate samples. It should be a super simple representation. And then you need various inference methods that can return that representation.

**Comment by jsalvatier** [comment]
I found the notebook a bit confusing, though I think I figured it out. 

Definitely call `histogram` something else. Its hard to think about this way. Maybe `posterior_samples` or `empirical_posterior`?.

I would put the section on optimization *before* the section showing that you can do regular inference (especially since that part seems to be standard variational inference?).

I would also try to make the Optimization section a bit clearer. For example I would explicitly call `y_` "`loss`" instead. It took me a while to think through what exactly `y_` was and what we were doing to it. 

You should also draw the minimum you found on a graph with the data and such.

**Comment by aseyboldt** [comment]
I don't like the name `Approximation` either. I think the problem is that the name tells us *how* something is stored (namely approximate) and not what it is.
Why not call `Approximation` `Posterior` then? The child classes would then represent different (approximate) ways to specify a posterior. Histogram could then be called `Sampled` (or maybe even `Trace`, but that might be confusing if we have the other trace type around), which would match the other names `MeanField` etc in that the name specifies how the posterior is specified. And probably also rename the module to `posterior`.
There wouldn't be a reason to put those in the global `pymc3` namespace, is there?

**Comment by jsalvatier** [comment]
Thoughts on `Optimizer`: 

I would prefer a stateless function rather than an object. State is generally bad. Makes things hard to reason about. 

The history storing callback should definitely be inside of `Optmizer`. 

Optimizer should probably also return the loss minimizing x, no? 

I want to support random search as optimizer because hyperparameter optimization often proceeds that   way.

**Comment by ferrine** [comment]
@aseyboldt `Posterior` is a bad name for base class for several reasons.

1. It is not actually true posterior
2. child classes names are not consistent
    compare  `class MeanField(Posterior)`  with `class MeanField(Approximation)`
    >  The child classes would then represent different (approximate) ways to specify a posterior
    
    If they are approximations they should be associated with approximations, not true posterior that we never know

3. 
    > rename the module to posterior.

    maybe it well be needed in future, when we decide to make unified inference result

@jsalvatier 
1.   > then you just need a common representation

     In my dreams, pymc3 API will be changed a lot

2. Comments about Jupiter notebook, when I return to this PR I'll bake it better

3. > State is generally bad
    
     Not always, it is better for fine tuning, monitoiring and debugging. If you are not desired with result there is no need to recompile step functions etc

What about renaming `Histogram`, I considered `Histogram->Empirical`

**Comment by twiecki** [comment]
`Empirical` seems to be the smallest common denominator. It's an empirical approximation from samples, which makes sense.

**Comment by fonnesbeck** [comment]
I'm on board with `Empirical`. 

We can talk about API changes on the next (first!) monthly PyMC video call.

**Comment by springcoil** [comment]
When's the monthly call?

**Comment by aseyboldt** [comment]
How about `ApproximatePosterior`or `PosteriorApproximation` then? I think the word Posterior should be in there somehow. I have to admit it took me an (embarrassingly) long time to understand what this whole module is about :-)

Empirical

> based on, concerned with, or verifiable by observation or experience rather than theory or pure logic.

How is a trace empirical? It's got nothing to do with observation or experience. The similarity to an empirical distribution (a distribution that is based on measurements) is really superficial.

I kind of like `Sampled`, but I'd prefer histogram over empirical.

On April 15, 2017 6:55:17 PM CEST, Maxim Kochurov <notifications@github.com> wrote:
>@aseyboldt `Posterior` is a bad name for base class for several
>reasons.
>
>1. It is not actually true posterior
>2. child classes names are not consistent
>compare  `class MeanField(Posterior)`  with `class
>MeanField(Approximation)`
>>  The child classes would then represent different (approximate) ways
>to specify a posterior
>    
>If they are approximations they should be associated with
>approximations, not true posterior that we never know
>
>3. 
>    > rename the module to posterior.
>
>maybe it well be needed in future, when we decide to make unified
>inference result
>
>@jsalvatier 
>1.   > then you just need a common representation
>
>     In my dreams, pymc3 API will be changed a lot
>
>2. Comments about Jupiter notebook, when I return to this PR I'll bake
>it better
>
>3. > State is generally bad
>    
>Not always, it is better for fine tuning, monitoiring and debugging. If
>you are not desired with result there is no need to recompile step
>functions etc
>
>What about renaming `Histogram`, I considered `Histogram->Empirical`
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/pymc-devs/pymc3/pull/1953#issuecomment-294305088

-Adrian

**Comment by fonnesbeck** [comment]
@aseyboldt its empirical in the sense that it is constructed from samples, rather than from a parametric function. i.e. the observed outcomes from a sampling procedure.

**Comment by twiecki** [comment]
@aseyboldt Do you find that convincing?

@ferrine I think we should change it to Empirical for now, definitely an improvement over Histogram.

**Comment by ferrine** [comment]
Name change is going to be in #2027

**Comment by aseyboldt** [comment]
@twiecki  I still think it's a bit strange to call draws from a posterior "empirical", but I don't feel that strongly about it. If you're all fine with it, then go ahead.

**Comment by fonnesbeck** [comment]
@aseyboldt I can see your point as well. What we have is analogous to a Dirichlet process, in that it represents some underlying distribution, but is constructed from a sample. What do we call that?

**Comment by junpenglao** [comment]
I still cannot really wrap my head around the application, so in this case you fitted the model and got the estimated parameters and then used them to find the function minima? It seems to be quite different from the usually application of Bayesian Optimaization. 
I think it would be great to have Bayesian Optimaization in PyMC3, a small wrapper taking the black box function and a pm.Model (e.g., a pymc3 GP model) as input, and find the function minima.

**Comment by ferrine** [comment]
Yea. Here is a simple wrapper to optimize such objectives but I do not like my API

**Comment by junpenglao** [comment]
@ferrine I think we can come back to this when the VI API is stabilised.

**Comment by ferrine** [comment]
Sure

**Comment by springcoil** [comment]
Any update on this now that opvi is stable?

**Comment by junpenglao** [comment]
@springcoil we should wait after https://github.com/pymc-devs/pymc3/pull/2416 is merged. Also maybe it is better to extend this into a project of its own (like GPflowOpt)?

**Comment by ferrine** [comment]
I like the idea. OPVI is wip again and comes with new cool features and speed as well. I don't plan to work on this pr this month.

**Comment by twiecki** [comment]
@ferrine I'm closing this as it seem stale, feel free to reopen if you disagree.

**Pull request closed without merging**